cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -1629,29 +1594,33 @@ private void handleProgressWithCompletion(ExecutionTask task, List<ExecutionTask
 
     /**
      * Periodically checks the metadata to see if leadership reassignment has finished or not.
+     * @param result Elect leaders result.
      */
-    private void waitForLeadershipTasksToFinish() {
+    private void waitForLeadershipTasksToFinish(ElectLeadersResult result) {
       List<ExecutionTask> finishedTasks = new ArrayList<>();
       Set<Long> stoppedTaskIds = new HashSet<>();
       Set<Long> deletedTaskIds = new HashSet<>();
       Set<Long> deadTaskIds = new HashSet<>();
+
+      // Process result to ensure acceptance of leader election request on broker-side and identify deleted tasks.
+      Set<TopicPartition> deletedUponSubmission = new HashSet<>();
+      ExecutionUtils.processElectLeadersResult(result, deletedUponSubmission);
+
+      boolean retry;
       do {
-        // If there is no finished tasks, we need to check if anything is blocked.
-        maybeReexecuteLeadershipTasks();
         Cluster cluster = getClusterForExecutionProgressCheck();
 
         List<ExecutionTask> slowTasksToReport = new ArrayList<>();
         boolean shouldReportSlowTasks = _time.milliseconds() - _lastSlowTaskReportingTimeMs > _slowTaskAlertingBackoffTimeMs;
         for (ExecutionTask task : inExecutionTasks()) {
           TopicPartition tp = task.proposal().topicPartition();
-          if (_stopSignal.get() == FORCE_STOP_EXECUTION) {
```

Comment:
This if statement was true only if the signal was `FORCE_STOP`.

Reply:
1. We made it evaluate to `true` for regular `STOP`, because we want to break this loop faster to make CC available for a new execution in case the user intends to stop an execution during a leadership movement.
2. Alternatively, we can completely drop this branch, but dropping it has the drawback of the `waitForLeadershipTasksToFinish()` taking longer.

Note that both options are eventually going to stop the execution as soon as `waitForLeadershipTasksToFinish()` completes, so the decision to keep or remove this branch will not cause an eventual correctness issue. Contrary to replica reassignments, there is no way to gracefully cancel an ongoing leadership reassignment. A leadership reassignment is expected to be a quick operation as there is no data move involved in the process.